tests/scripts/thread-cert/Cert_6_1_02_REEDAttach.py

```diff
@@ -0,0 +1,377 @@
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+import unittest
+import copy
+
+import thread_cert
+import config
+from pktverify.consts import WPAN_DATA_REQUEST, MLE_PARENT_REQUEST, MLE_PARENT_RESPONSE, MLE_CHILD_UPDATE_REQUEST, MLE_CHILD_UPDATE_RESPONSE, MLE_CHILD_ID_REQUEST, MLE_CHILD_ID_RESPONSE, ADDR_SOL_URI, SOURCE_ADDRESS_TLV, MODE_TLV, TIMEOUT_TLV, CHALLENGE_TLV, RESPONSE_TLV, LINK_LAYER_FRAME_COUNTER_TLV, MLE_FRAME_COUNTER_TLV, ROUTE64_TLV, ADDRESS16_TLV, LEADER_DATA_TLV, NETWORK_DATA_TLV, TLV_REQUEST_TLV, SCAN_MASK_TLV, CONNECTIVITY_TLV, LINK_MARGIN_TLV, VERSION_TLV, ADDRESS_REGISTRATION_TLV, NL_MAC_EXTENDED_ADDRESS_TLV, NL_RLOC16_TLV, NL_STATUS_TLV, NL_ROUTER_MASK_TLV, COAP_CODE_ACK
+from pktverify.packet_verifier import PacketVerifier
+from pktverify.null_field import nullField
+
+LEADER = 1
+REED = 2
+MTD = 3
+
+# Test Purpose and Description:
+# -----------------------------
+# The purpose of this test case is to validate that the DUT is able to successfully
+# attach to a network
+#
+# Test Topology:
+# -------------
+#  Leader
+#    |
+#  REED
+#    |
+#   DUT
+#
+# DUT Types:
+# ----------
+#  ED
+#  SED
+
+
+class Cert_6_1_2_REEDAttach_Base(thread_cert.TestCase):
+    USE_MESSAGE_FACTORY = False
+
+    TOPOLOGY = {
+        LEADER: {
+            'name': 'LEADER',
+            'mode': 'rdn',
+            'panid': 0xface,
+            'allowlist': [REED]
+        },
+        REED: {
+            'name': 'REED',
+            'mode': 'rdn',
+            'panid': 0xface,
+            'router_upgrade_threshold': 0,
+            'allowlist': [LEADER, MTD]
+        },
+        MTD: {
+            'name': 'DUT',
+            'is_mtd': True,
+            'panid': 0xface,
+            'timeout': config.DEFAULT_CHILD_TIMEOUT,
+            'allowlist': [REED]
+        },
+    }
+
+    def test(self):
+        self.nodes[LEADER].start()
+        self.simulator.go(5)
+        self.assertEqual(self.nodes[LEADER].get_state(), 'leader')
+
+        self.nodes[REED].start()
+        self.simulator.go(config.MAX_ADVERTISEMENT_INTERVAL)
+        self.assertEqual(self.nodes[REED].get_state(), 'child')
+        self.collect_rloc16s()
+
+        self.nodes[MTD].start()
+        self.simulator.go(5)
+        self.assertEqual(self.nodes[MTD].get_state(), 'child')
+        self.assertEqual(self.nodes[REED].get_state(), 'router')
+        self.collect_ipaddrs()
+        self.collect_rlocs()
+
+        self.simulator.go(config.DEFAULT_CHILD_TIMEOUT)
+
+        dut_addr = self.nodes[MTD].get_ip6_address(config.ADDRESS_TYPE.LINK_LOCAL)
+        self.assertTrue(self.nodes[REED].ping(dut_addr))
```

Comment:
I may suggest we keep simply following the TP,  in which
> Step 10.  The ping is from Leader to DUT for ED DUT.
> Step 12. The ping is from REED to DUT for SED DUT.

Reply:
Noticing there should be issue in the TP as Leader should not be able to ping DUT via link local address for Step 10 in the given topology.

Fine to me to keep ping from REED to DUT via link local address so far, but would you please file a Thread Group ticket to either update the Device in Step 10 from Leader to REED, or change the address type to non-link local address in the description?

Reply:
sure